mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManagerV2.java

```diff
@@ -0,0 +1,360 @@
+    }
+
+    @Override
+    public Mono<Void> removeGlobalMaxMessageReactive() {
+        return cassandraQuotaLimitDao.deleteQuotaLimit(QuotaLimitKey.of(QuotaComponent.MAILBOX, QuotaScope.GLOBAL, GLOBAL_IDENTIFIER, QuotaType.COUNT));
+    }
+
+    @Override
+    public Optional<QuotaSizeLimit> getGlobalMaxStorage() {
+        return getGlobalMaxStorageReactive().blockOptional();
+    }
+
+    @Override
+    public Mono<QuotaSizeLimit> getGlobalMaxStorageReactive() {
+        return getMaxStorageReactive(QuotaScope.GLOBAL, GLOBAL_IDENTIFIER);
+    }
+
+    @Override
+    public Optional<QuotaCountLimit> getGlobalMaxMessage() {
+        return getGlobalMaxMessageReactive().blockOptional();
+    }
+
+    @Override
+    public Mono<QuotaCountLimit> getGlobalMaxMessageReactive() {
+        return getMaxMessageReactive(QuotaScope.GLOBAL, GLOBAL_IDENTIFIER);
+    }
+
+    @Override
+    public Map<Quota.Scope, QuotaCountLimit> listMaxMessagesDetails(QuotaRoot quotaRoot) {
+        return listMaxMessagesDetailsReactive(quotaRoot).block();
+    }
+
+    @Override
+    public Mono<Map<Quota.Scope, QuotaCountLimit>> listMaxMessagesDetailsReactive(QuotaRoot quotaRoot) {
+        return Flux.merge(
+            getMaxMessageReactive(QuotaScope.USER, quotaRoot.getValue())
+                .map(limit -> Pair.of(Quota.Scope.User, limit)),
+            Mono.justOrEmpty(quotaRoot.getDomain())
+                .flatMap(domain -> getMaxMessageReactive(QuotaScope.DOMAIN, domain.asString()))
+                .map(limit -> Pair.of(Quota.Scope.Domain, limit)),
+            getGlobalMaxMessageReactive()
+                .map(limit -> Pair.of(Quota.Scope.Global, limit)))
+            .collect(ImmutableMap.toImmutableMap(
+                Pair::getKey,
+                Pair::getValue));
+    }
+
+    @Override
+    public Map<Quota.Scope, QuotaSizeLimit> listMaxStorageDetails(QuotaRoot quotaRoot) {
+        return listMaxStorageDetailsReactive(quotaRoot).block();
+    }
+
+    @Override
+    public Mono<Map<Quota.Scope, QuotaSizeLimit>> listMaxStorageDetailsReactive(QuotaRoot quotaRoot) {
+        return Flux.merge(
+            getMaxStorageReactive(QuotaScope.USER, quotaRoot.getValue())
+                .map(limit -> Pair.of(Quota.Scope.User, limit)),
+            Mono.justOrEmpty(quotaRoot.getDomain())
+                .flatMap(domain -> getMaxStorageReactive(QuotaScope.DOMAIN, domain.asString()))
+                .map(limit -> Pair.of(Quota.Scope.Domain, limit)),
+            getGlobalMaxStorageReactive()
+                .map(limit -> Pair.of(Quota.Scope.Global, limit)))
+            .collect(ImmutableMap.toImmutableMap(
+                Pair::getKey,
+                Pair::getValue));
+    }
+
+    @Override
+    public QuotaDetails quotaDetails(QuotaRoot quotaRoot) {
+        return quotaDetailsReactive(quotaRoot)
+            .block();
+    }
+
+    @Override
+    public Mono<QuotaDetails> quotaDetailsReactive(QuotaRoot quotaRoot) {
```

Comment:
Please let's try to reduce the amount of reads done here.

One read for user limits (size and count), one read for domain limit (size and count) and one read for global limit (size and count)

We should write a test for this with StatementRecoder and TestingSession with Cassandra tests, in order to ensure that we minimize request to the DB.

This is important for performance.

Reply:
Ok apparently done, great!